doc/notes/11702_attr_order_int.adoc

```diff
@@ -0,0 +1,115 @@
+This note documents the implementation of attribute order retention in the
+`xtuml2masl` flow.
+
+== 2 Introduction and Background
+
+A user has an existing project where there is a need to migrate data stored
+in a persistence database from an existing domain to a new version of that
+domain.
+
+This is accomplished by dumping the database as SQL which provides the
+create and insert statements for the database schema and contents.
+This dumped data is then manipulated to make changes to the schema and
+content, e.g. create/delete new classes, relationships, attributes, etc.
+
+Unfortunately this is proving more difficult than in the past as BridgePoint
+is not dumping the class definitions in the '.mod' file the same way as the
+dumper for iUML was.
+
+In the '.mod' files dumped from iUML the attributes were in the order they
+appeared on the class diagram.  In the '.mod' files from BridgePoint the
+class identifiers appear to be always the first attributes defined, and it
+does not follow the order of the class diagram.
+
+Therefore when the domain is built, and the schema is created, it is
+different from the schema of the previous iUML version and migrating
+the data from an existing system is problematic.
+
+== 3 Requirements
+
+=== 3.1 Attribute Ordering
+The `xtuml2masl` process shall produce a '.mod' file with attributes ordered
+in the class definitions in the same sequence as maintained in the BridgePoint
+xtUML editor.
+
+== 4 Work Required
+
+=== 4.1 `x2m`
+The action language in the `x2m` application (maslout) was performing a
+`select many` and `for each` with no ordering.  This is to be changed to
+select the first attribute in the sequence of attributes related to a class.
+Iteration then needs to select across R103 to the next attribute.
+
+=== 4.2 `masl`
+The model of MASL was not maintaining the order of the attributes.  A
+reflexive ordering association is added the attribute class.
+
+`masl` deals with forward references by creating elements referred to by
+elements being created ahead of them.  These "anticipatory" creations get
+augmented when the primary creation command comes.  It will be necessary
+to do ordering at primary creation time rather than anticipatory creation
+time.
+
+=== 4.3 Serial MASL Specification
+Update the SMASL spec document to add parameter to attribute command to
+include the name of the previous attribute.
+
+== 5 Implementation Comments
+
+=== 5.1 Architecture Ordering
+Both `m2x` and `masl` are rendered as C source code by MC-3020.
+Architecturally, instance collections are implemented with singly linked
+lists.  It was observed that after sequencing attributes in modeled
+order in `m2x` that the headline issue was effectively solved.  `masl`
+rendered the attributes in the order created, and so the ordering was
+"accidentally" maintained.  A dependence upon the underlying architectural
+collection mechanism is not desirable, so `masl` was outfitted with a
+modeled ordering association.
+
```

Comment:
I also note here that as soon as you start editing the model and change the order of the attributes, or insert new attributes, things will get out of order... The dependence on the architectural mechanism only works with round trip without significant editing in between import and export.